Files can now be deleted

s1050613 opened the pull request:
Have no idea what I’m doing. I’ve looked at other similar bits of code and I hope that this is how it’s meant to be done, since I have no understanding on how PyAngelo’s back-end functions.

This should _(should)_ allow users to delete files on their sketches, apart from the main.py file. It will give them a confirmation warning (thx Habbison) and the delete buttons are nicely styled (at least they are for me).

Probably got some things wrong, forgot some things or made some typos, so please check over because I have most likely done something wrong somewhere!

s1050613 committed:
Update routes.php

I have no idea what I am doing. Ah well.

s1050613 committed:
Update services.php

So um I’m trying to add the ability for files to be deleted. Still have no idea what I’m doing.

s1050613 committed:
Create SketchDeleteFileController.php

Ok now gotta change sketchRepository and sketchFiles..

s1050613 committed:
Update SketchRepository.php

s1050613 committed:
Update SketchFiles.php

s1050613 committed:
Update editor.js

(hopefully) (if all works well) (maybe)
I have tried to make it look as similar as the rest of the code so it won’t stand out.

s1050613 committed:
Added styling for the remove file buttons

Arghh gonna be tricky to test the styling—let’s guess for now.

s1050613 committed:
Update sketch.css

LOL the styling looked TERRIBLE; I visualised the buttons yellow instead of dark purple. This looks a bit better.

s1050613 committed:
Update editor.js

Just realised… oops!
Also changed the multiplication symbol to the HTML entity code since weird happenings were happening.

s1050613 committed:
Update sketch.css

Silly me!

s1050613 committed:
Added warning when deleting files

We don’t want people accidentally deleting files, do we?!
Thanks to Habbison for the idea.

s1050613 committed:
You can now check if files exist

We don’t want people deleting inexistent files do we?!

s1050613 committed:
Update SketchFiles.php

Arghh typos

s1050613 committed:
You can’t delete main.py or inexistent files

And fixed a typo!
Lol that last if statement looks a bit strange…

pingskills commented on src/PyAngelo/Repositories/SketchRepository.php:
```diff
@@ -18,6 +18,8 @@ public function getSketchFiles($sketchId);
 
   public function addSketchFile($sketchId, $filename);
+  
+  public function deleteSketchFile($sketchId, $filename);
```
This is an abstract method. You must implement this in MysqlSketchRepository.php where it must issue a delete statement using SQL. Notice how each method in this file is implemented in MysqlSketchRepository.php?

s1050613 replied:
Ahh! What a giant rabbit hole this is!
I’ve put it in there and I’m pretty sure that’s how it’s done.

pingskills commented on resources/assets/js/editor.js:
```diff
@@ -148,7 +148,21 @@ function setupEditor(response) {
+      let deleteButton = document.createElement('span');
+      deleteButton.innerHTML = '&times;';
+      deleteButton.onclick = function(this) {
+        let span = this.parentNode;
```
span is already defined in this context so we don't need this line.

s1050613 replied:
Not very familiar with ES6 variables; thank you!

pingskills commented on resources/assets/js/editor.js:
```diff
@@ -148,7 +148,21 @@ function setupEditor(response) {
+    if(file.filename != 'main.py') {
+      let deleteButton = document.createElement('span');
+      deleteButton.innerHTML = '&times;';
+      deleteButton.onclick = function(this) {
```
The "this" parameter is causing an error for me. Should this be event?

s1050613 replied:
Oops! Don’t know why I ever did that..
Just goes to show that however confident you are in your code, you should always test it!
Thanks; fixed.

pingskills commented on config/services.php:
```diff
@@ -669,6 +669,16 @@
 });
 
+$di->set('SketchDeleteFileController', function () use ($di) {
+  return new Pyangelo\Controllers\Sketch\SketchDeleteFileController (
```
Pyangelo needs to be "PyAngelo"

s1050613 replied:
Thank you; fixed!

pingskills reviewed (commented):
Thanks for submitting this pull request. A lot of it looks good! Have you created an environment where you can test these changes? I have made some comments against the code that need to be fixed. I tried modifying some of these but there are still some errors that need to be fixed. There is also a series of tests that should be written for the code that will help avoid these errors. I'm impressed with what you have created. Well done.

s1050613 committed:
Merge branch 'NocturnalRage:master' into master

s1050613 committed:
Update services.php

s1050613 committed:
Update editor.js

Doesn’t look like much, but I’ve tested and tested with different scenario promises that would returned by the HTTP request. I unfortunately don’t have a proper testing environment, but I’ve tested the fetch being rejected, the fetch resolving but with an error HTTP status code, the fetch resolving with a valid HTTP status code but the response.json() failing, and everything working fine. It also will give an alert if the HTTP request returns a filename which doesn’t exist, which I don’t think should ever happen.

Thank you for reviewing my pull request! I will investigate more and try to fix everything up.

s1050613 committed:
Update MysqlSketchRepository.php

Back-end is definitely not my forte!
From what I’ve inferred from the rest of the code, the bind_param adds the values to the SQL statement without the risk of a SQL injection, with “is” meaning “integer string” for the types of the sketch_id and filename variables.
Then, I’ve Googled a bit and from what I’ve gathered is the insert_id and $fileId variable is telling what number sketch it was, but because we’re deleting a row, it won’t return anything so we’re making it return whether or not the PHP statement worked.

Probably done something wrong or made a few typos as usual; I’ve checked over it but I’m not too sure.

s1050613 committed:
Update MysqlSketchRepository.php

Oops! Put tabs instead of spaces; looked ugly while looking at it so changed.

s1050613 committed:
Update editor.js

Fixed some more ugly tabbing

s1050613 committed:
Merge branch 'NocturnalRage:master' into master

pingskills committed:
Update SketchDeleteFileController.php

pingskills committed:
Update MysqlSketchRepository.php

pingskills committed:
Update SketchFiles.php

pingskills reviewed (commented):
Hey Julian, this is looking close now. There are still a few things to update:
1) can you create the file views/sketch/delete.json.php as a copy of add.json.php - It's good to have a different view here in case we ever need something different.
2) Copy tests/views/sketch/SketchAddJsonTest.php to SketchDeleteJsonTest.php and modify accordingly
3)Copy tests/src/PyAngelo/Controllers/Sketch/SketchAddFileController.php to SketchDeleteFileController.php and modify accordingly
4) Can you merge all the commits so far to a single commit?

s1050613 commented:
Hello Mr. Plumb, thank you for reviewing this pull request. I never thought that one day I would be contributing to PyAngelo! Thank you for fixing up my errors. I will follow your list and hopefully I’ll get them done soon!

s1050613 committed:
Create delete.json.php

Should be that!
Step 1 done!

s1050613 committed:
Create SketchDeleteJsonTest.php

Ahh so this is the testing!
Believe that’s how it’s done.

s1050613 committed:
Create SketchDeleteFileControllerTest.php

From what I’ve gathered, this is where you’ll be testing it in your environment. Since when deleting files it needs to check that a) it isn’t main.py and b) the file to be deleted exists, I’ve added some methods to test those out as well, but I’ve probably done something wrong along the way there as well.

I’ve copy-pasted the error messages from SketchDeleteFileController.php so they match up. From what I believe, you’ll call these methods individually to test the error messages in SketchDeleteFileController.php. I don’t think that I need to change much in these testing methods? I’m not sure how it operates.

Probably some typos in there.

s1050613 commented:
I’ve Googled around a bit; don’t think I can merge the commits into one without write access.

pingskills commented:
Nice work. I'll check out the updates this week. Don't worry about merging the commits for this request. I can do it on my end when I merge the pull request.

pingskills committed:
Update SketchDeleteFileControllerTest.php

pingskills committed:
Update SketchDeleteJsonTest.php

pingskills committed:
Update SketchDeleteFileController.php

pingskills committed:
Update editor.js

pingskills committed:
Update sketch.css

s1050613 commented:
Wow! Thank you so much Mr. Plumb!

Thank you for helping me out with this experience. I have gained an insight into how real websites function, and it’s harder than I thought!

Thank you again for guiding me through this.

s1050613 commented:
Wow! Thank you so much Mr. Plumb!

Thank you for helping me out with this experience. I have gained an insight into how real websites function, and it’s harder than I thought!

Thank you again for guiding me through this.

pingskills commented:
You're welcome. I'm about to upload it to the live site :) Great job! I was really impressed with your efforts. I'm going to create a few issues on github so if you want to have a go at those feel free. Or if you don't want to that is fine too!

pingskills commented:
The changes are now live on PyAngelo.com. Have a play and let me know if you find something that doesn't work as expected. Thanks again for your help! An amazing job!!!!

s1050613 commented:
Thank you! It seems to be working very smoothly.

Thanks for the issues; you have given me more afternoon entertainment!
